mockgcp/common/fields/updatemask.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fields
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+func UpdateByFieldMask(old, new proto.Message, updatePaths []string) error {
+	var errs []error
+	for _, path := range updatePaths {
+		if err := walk(old, new, path); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+func walk(old, new proto.Message, path string) error {
+	segments := strings.Split(path, ".")
+	if len(segments) == 1 {
+		return replace(old, new, segments[0])
+	}
+
+	subPath := strings.TrimPrefix(path, segments[0]+".")
```

Comment:
I like this "iterative" approach to parsing the field path, because it might one day be able to cope with more complex paths (e.g. `foo.bars[1]`).  👍 

You might consider `subPath := strings.Join(segments[1:], ".")`.

Or once I saw that, I realized you can probably also do this:

```
segments := strings.SplitN(path, ".", 2)
if len(segments) == 1 {
		return replace(old, new, segments[0])
	}

subPath := segments[1]
```

Reply:
Thank you! I like it.